File: truss-compiler-maven-plugin/src/main/java/truss/compiler/generator/syntax/GenerateSyntax.java
package truss.compiler.generator.syntax;

import truss.compiler.generator.CodeWriter;
import truss.compiler.generator.syntax.xml.SyntaxClass;
import truss.compiler.generator.syntax.xml.SyntaxEnum;
import truss.compiler.generator.syntax.xml.SyntaxProperty;
import truss.compiler.generator.syntax.xml.SyntaxSchema;
import org.apache.commons.io.IOUtils;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import java.io.*;
import java.util.*;

public class GenerateSyntax {
    private final Map<String, SyntaxClass> classes = new HashMap<>();

    public void generate(File definition, File outputDirectory) throws JAXBException, IOException {
        SyntaxSchema schema = (SyntaxSchema)JAXBContext.newInstance(SyntaxSchema.class)
            .createUnmarshaller()
            .unmarshal(definition);

        Collections.sort(schema.getClasses(), new Comparator<SyntaxClass>() {
            @Override
            public int compare(SyntaxClass a, SyntaxClass b) {
                return a.getName().compareTo(b.getName());
            }
        });

        outputDirectory = new File(outputDirectory, schema.getPackageName().replace('.', File.separatorChar));

        for (SyntaxClass syntaxClass : schema.getClasses()) {
            classes.put(syntaxClass.getName(), syntaxClass);
        }

        for (SyntaxClass syntaxClass : schema.getClasses()) {
            if (!syntaxClass.isIgnore()) {
                generateClass(outputDirectory, syntaxClass, schema.getPackageName());
            }
        }

        for (SyntaxEnum syntaxEnum : schema.getEnums()) {
            generateEnum(outputDirectory, syntaxEnum, schema.getPackageName());
        }

        generateVisitor(schema, outputDirectory);
        generateActionVisitor(schema, outputDirectory);
        generateAbstractVisitor(schema, outputDirectory);
        generateAbstractActionVisitor(schema, outputDirectory);
        generateWalker(schema, outputDirectory);
        generateSyntaxKind(schema, outputDirectory);
    }

    private void generateVisitor(SyntaxSchema schema, File outputDirectory) throws IOException {
        CodeWriter cw = new CodeWriter();

        cw.writeln("package %s;", schema.getPackageName());
        cw.writeln();

        cw.writeln("public interface SyntaxVisitor {");
        cw.indent();

        cw.writeln("boolean isDone();");

        for (SyntaxClass syntaxClass : schema.getClasses()) {
            if (syntaxClass.isAbstract()) {
                continue;
            }

            cw.writeln();
            cw.writeln("void visit%s(%s syntax) throws Exception;", getStrippedName(syntaxClass), syntaxClass.getName());
        }

        cw.unIndent();
        cw.writeln("}");

        writeWhenChanged(new File(outputDirectory, "SyntaxVisitor.java"), cw.toString());
    }

    private void generateActionVisitor(SyntaxSchema schema, File outputDirectory) throws IOException {
        CodeWriter cw = new CodeWriter();

        cw.writeln("package %s;", schema.getPackageName());
        cw.writeln();

        cw.writeln("public interface SyntaxActionVisitor<T> {");
        cw.indent();

        boolean hadOne = false;

        for (SyntaxClass syntaxClass : schema.getClasses()) {
            if (syntaxClass.isAbstract()) {
                continue;
            }

            if (hadOne) {
                cw.writeln();
            } else {
                hadOne = true;
            }
            cw.writeln("T visit%s(%s syntax) throws Exception;", getStrippedName(syntaxClass), syntaxClass.getName());
        }

        cw.unIndent();
        cw.writeln("}");

        writeWhenChanged(new File(outputDirectory, "SyntaxActionVisitor.java"), cw.toString());
    }

    private void generateAbstractVisitor(SyntaxSchema schema, File outputDirectory) throws IOException {
        CodeWriter cw = new CodeWriter();

        cw.writeln("package %s;", schema.getPackageName());
        cw.writeln();

        cw.writeln("public class AbstractSyntaxVisitor implements SyntaxVisitor {");
        cw.indent();

        cw.writeln("private boolean done;");
        cw.writeln();

        cw.writeln("public boolean isDone() {");
        cw.indent();
        cw.writeln("return done;");
        cw.unIndent();
        cw.writeln("}");
        cw.writeln();
        cw.writeln("public void setDone(boolean done) {");
        cw.indent();
        cw.writeln("this.done = done;");
        cw.unIndent();
        cw.writeln("}");
        cw.writeln();
        cw.writeln("public void defaultVisit(SyntaxNode syntax) throws Exception {");
        cw.writeln("}");

        for (SyntaxClass syntaxClass : schema.getClasses()) {
            if (syntaxClass.isAbstract()) {
                continue;
            }

            cw.writeln();
            cw.writeln("@Override");
            cw.writeln("public void visit%s(%s syntax) throws Exception {", getStrippedName(syntaxClass), syntaxClass.getName());
            cw.indent();
            cw.writeln("defaultVisit(syntax);");
            cw.unIndent();
            cw.writeln("}");
        }

        cw.unIndent();
        cw.writeln("}");

        writeWhenChanged(new File(outputDirectory, "AbstractSyntaxVisitor.java"), cw.toString());
    }

    private void generateAbstractActionVisitor(SyntaxSchema schema, File outputDirectory) throws IOException {
        CodeWriter cw = new CodeWriter();

        cw.writeln("package %s;", schema.getPackageName());
        cw.writeln();

        cw.writeln("public class AbstractSyntaxActionVisitor<T> implements SyntaxActionVisitor<T> {");
        cw.indent();

        cw.writeln("public T defaultVisit(SyntaxNode syntax) throws Exception {");
        cw.indent();
        cw.writeln("return null;");
        cw.unIndent();
        cw.writeln("}");

        boolean hadOne = false;

        for (SyntaxClass syntaxClass : schema.getClasses()) {
            if (syntaxClass.isAbstract()) {
                continue;
            }

            if (hadOne) {
                cw.writeln();
            } else {
                hadOne = true;
            }
            cw.writeln("@Override");
            cw.writeln("public T visit%s(%s syntax) throws Exception {", getStrippedName(syntaxClass), syntaxClass.getName());
            cw.indent();
            cw.writeln("return defaultVisit(syntax);");
            cw.unIndent();
            cw.writeln("}");
        }

        cw.unIndent();
        cw.writeln("}");

        writeWhenChanged(new File(outputDirectory, "AbstractSyntaxActionVisitor.java"), cw.toString());
    }

    private void generateWalker(SyntaxSchema schema, File outputDirectory) throws IOException {
        CodeWriter cw = new CodeWriter();

        cw.writeln("package %s;", schema.getPackageName());
        cw.writeln();
        cw.writeln("import truss.compiler.support.ImmutableArray;");
        cw.writeln();

        cw.writeln("public class SyntaxTreeWalker implements SyntaxVisitor {");
        cw.indent();

        cw.writeln("private boolean done;");
        cw.writeln();

        cw.writeln("public boolean isDone() {");
        cw.indent();
        cw.writeln("return done;");
        cw.unIndent();
        cw.writeln("}");
        cw.writeln();
        cw.writeln("public void setDone(boolean done) {");
        cw.indent();
        cw.writeln("this.done = done;");
        cw.unIndent();
        cw.writeln("}");
        cw.writeln();
        cw.writeln("public void visitList(ImmutableArray<? extends SyntaxNode> list) throws Exception {");
        cw.indent();
        cw.writeln("for (SyntaxNode node : list) {");
        cw.indent();
        cw.writeln("visit(node);");
        cw.unIndent();
        cw.writeln("}");
        cw.unIndent();
        cw.writeln("}");
        cw.writeln();
        cw.writeln("public void visit(SyntaxNode syntax) throws Exception {");
        cw.indent();
        cw.writeln("if (syntax != null) {");
        cw.indent();
        cw.writeln("syntax.accept(this);");
        cw.unIndent();
        cw.writeln("}");
        cw.unIndent();
        cw.writeln("}");

        for (SyntaxClass syntaxClass : schema.getClasses()) {
            if (syntaxClass.isAbstract()) {
                continue;
            }

            cw.writeln();
            cw.writeln("@Override");
            cw.writeln("public void visit%s(%s syntax) throws Exception {", getStrippedName(syntaxClass), syntaxClass.getName());
            cw.indent();

            for (SyntaxProperty property : resolveParameters(syntaxClass)) {
                if (!isSyntaxNode(property)) {
                    continue;
                }

                if (property.isList()) {
                    cw.writeln("visitList(syntax.%s());", getMethodName(property));
                } else {
                    cw.writeln("visit(syntax.%s());", getMethodName(property));
                }
            }

            cw.unIndent();
            cw.writeln("}");
        }

        cw.unIndent();
        cw.writeln("}");

        writeWhenChanged(new File(outputDirectory, "SyntaxTreeWalker.java"), cw.toString());
    }

    private void generateSyntaxKind(SyntaxSchema schema, File outputDirectory) throws IOException {
        CodeWriter cw = new CodeWriter();

        cw.writeln("package %s;", schema.getPackageName());
        cw.writeln();

        cw.writeln("public enum SyntaxKind {");
        cw.indent();

        for (SyntaxClass syntaxClass : schema.getClasses()) {
            if (!syntaxClass.isAbstract()) {
                cw.writeln("%s,", getSyntaxKind(syntaxClass));
            }
        }

        cw.unIndent();
        cw.writeln("}");

        writeWhenChanged(new File(outputDirectory, "SyntaxKind.java"), cw.toString());
    }

    private boolean isSyntaxNode(SyntaxProperty property) {
        return property.getType().endsWith("Syntax");
    }

    private void generateClass(File outputDirectory, SyntaxClass schemaClass, String packageName) throws IOException {
        CodeWriter cw = new CodeWriter();

        cw.writeln("package %s;", packageName);
        cw.writeln();

        cw.writeln("import truss.compiler.*;");
        cw.writeln("import truss.compiler.symbols.*;");
        cw.writeln("import truss.compiler.syntax.*;");
        cw.writeln("import org.apache.commons.lang.Validate;");
        cw.writeln();
        cw.writeln("import truss.compiler.support.ImmutableArray;");
        cw.writeln();

        cw.writeln(
            "public %sclass %s extends %s {",
            schemaClass.isAbstract() ? "abstract " : "",
            schemaClass.getName(),
            schemaClass.getBase()
        );

        cw.indent();

        for (SyntaxProperty property : schemaClass.getProperties()) {
            cw.writeln("private final %s %s;", getType(property), getLocalName(property));
        }

        List<SyntaxProperty> parameters = resolveParameters(schemaClass);

        cw.writeln();
        cw.write(
            "%s %s(",
            schemaClass.isAbstract() ? "protected" : "public",
            schemaClass.getName()
        );

        writeParameters(cw, parameters);

        cw.writeln(") {");

        cw.indent();

        cw.write("super(");

        boolean hadOne = false;
        for (SyntaxProperty parameter : parameters) {
            if (schemaClass.getProperties().indexOf(parameter) != -1) {
                continue;
            }

            if (hadOne) {
                cw.write(", ");
            } else {
                hadOne = true;
            }

            cw.write(getLocalName(parameter));
        }

        cw.writeln(");");
        cw.writeln();

        for (SyntaxProperty parameter : parameters) {
            if (schemaClass.getProperties().indexOf(parameter) == -1) {
                continue;
            }

            if (!parameter.isNullable() && !isValueType(parameter.getType())) {
                cw.writeln("Validate.notNull(%s, \"%s\");", getLocalName(parameter), getLocalName(parameter));
            }
        }

        if (schemaClass.getValidation() != null) {
            cw.writeRaw(schemaClass.getValidation());
            cw.writeRaw("\n");
        }

        cw.writeln();

        for (SyntaxProperty property : schemaClass.getProperties()) {
            cw.writeln("this.%s = %s;", getLocalName(property), getLocalName(property));
        }

        cw.unIndent();
        cw.writeln("}");

        for (SyntaxProperty property : schemaClass.getProperties()) {
            cw.writeln();
            cw.writeln("public %s %s() {", getType(property), getMethodName(property));
            cw.indent();
            cw.writeln("return %s;", getLocalName(property));
            cw.unIndent();
            cw.writeln("}");
        }

        if (!schemaClass.isAbstract()) {
            cw.writeln();
            cw.writeln("@Override");
            cw.writeln("public SyntaxKind getKind() {");
            cw.indent();
            cw.writeln("return SyntaxKind.%s;", getSyntaxKind(schemaClass));
            cw.unIndent();
            cw.writeln("}");

            cw.writeln();
            cw.writeln("@Override");
            cw.writeln("public void accept(SyntaxVisitor visitor) throws Exception {");
            cw.indent();
            cw.writeln("if (!visitor.isDone()) {");
            cw.indent();
            cw.writeln("visitor.visit%s(this);", getStrippedName(schemaClass));
            cw.unIndent();
            cw.writeln("}");
            cw.unIndent();
            cw.writeln("}");

            cw.writeln();
            cw.writeln("@Override");
            cw.writeln("public <T> T accept(SyntaxActionVisitor<T> visitor) throws Exception {");
            cw.indent();
            cw.writeln("return visitor.visit%s(this);", getStrippedName(schemaClass));
            cw.unIndent();
            cw.writeln("}");
        }

        if (schemaClass.getMembers() != null) {
            cw.writeRaw(schemaClass.getMembers());
            cw.writeRaw("\n");
        }

        cw.unIndent();
        cw.writeln("}");

        writeWhenChanged(new File(outputDirectory, schemaClass.getName() + ".java"), cw.toString());
    }

    private void generateEnum(File outputDirectory, SyntaxEnum schemaEnum, String packageName) throws IOException {
        CodeWriter cw = new CodeWriter();

        cw.writeln("package %s;", packageName);
        cw.writeln();

        cw.writeln("import truss.compiler.*;");
        cw.writeln("import truss.compiler.symbols.*;");
        cw.writeln("import truss.compiler.syntax.*;");
        cw.writeln("import org.apache.commons.lang.Validate;");
        cw.writeln();
        cw.writeln("import truss.compiler.support.ImmutableArray;");
        cw.writeln();

        cw.writeln(
            "public enum %s {",
            schemaEnum.getName()
        );

        cw.indent();

        for (SyntaxProperty property : schemaEnum.getProperties()) {
            cw.writeln("%s,", property.getName());
        }

        cw.unIndent();
        cw.writeln("}");

        writeWhenChanged(new File(outputDirectory, schemaEnum.getName() + ".java"), cw.toString());
    }

    private String getSyntaxKind(SyntaxClass schemaClass) {
        String name = getStrippedName(schemaClass);

        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < name.length(); i++) {
            char c = name.charAt(i);

            if (i > 0 && Character.isUpperCase(c)) {
                sb.append('_');
            }
            sb.append(Character.toUpperCase(c));
        }

        return sb.toString();
    }

    private void writeParameters(CodeWriter cw, List<SyntaxProperty> parameters) {
        boolean hadOne = false;

        for (SyntaxProperty parameter : parameters) {
            if (hadOne) {
                cw.write(", ");
            } else {
                hadOne = true;
            }

            cw.write("%s %s", getType(parameter), getLocalName(parameter));
        }
    }

    private String getMethodName(SyntaxProperty property) {
        String methodName = property.getName();
        if (methodName.startsWith("Is") || methodName.startsWith("Has")) {
            methodName = getLocalName(property);
        } else {
            methodName = "get" + methodName;
        }
        return methodName;
    }

    private String getStrippedName(SyntaxClass syntaxClass) {
        final String postfix = "Syntax";
        assert syntaxClass.getName().endsWith(postfix);
        return syntaxClass.getName().substring(0, syntaxClass.getName().length() - postfix.length());
    }

    private boolean isValueType(String type) {
        switch (type) {
            case "boolean": return true;
            default: return false;
        }
    }

    private String getLocalName(SyntaxProperty property) {
        String name = property.getName();

        name = name.substring(0, 1).toLowerCase() + name.substring(1);

        switch (name) {
            case "finally":
            case "else":
            case "default":
                name = name + "_";
                break;
        }

        return name;
    }

    private String getType(SyntaxProperty property) {
        if (property.isList()) {
            return "ImmutableArray<" + property.getType() + ">";
        }

        return property.getType();
    }

    private List<SyntaxProperty> resolveParameters(SyntaxClass SyntaxClass) {
        List<SyntaxProperty> baseParameters;

        if (SyntaxClass.getBase() != null) {
            baseParameters = resolveParameters(classes.get(SyntaxClass.getBase()));
        } else {
            baseParameters = Collections.emptyList();
        }

        List<SyntaxProperty> parameters = new ArrayList<>();
        boolean hadLast = false;

        for (SyntaxProperty baseParameter : baseParameters) {
            if (!hadLast && baseParameter.isLast()) {
                hadLast = true;
                parameters.addAll(SyntaxClass.getProperties());
            }

            parameters.add(baseParameter);
        }

        if (!hadLast) {
            parameters.addAll(SyntaxClass.getProperties());
        }

        return parameters;
    }

    private void writeWhenChanged(File file, String content) throws IOException {
        if (file.exists()) {
            try (InputStream is = new FileInputStream(file)) {
                String current = IOUtils.toString(is);
                if (current.equals(content)) {
                    return;
                }
            }
        }

        file.getParentFile().mkdirs();

        try (OutputStream os = new FileOutputStream(file)) {
            IOUtils.write(content, os);
        }
    }
}

File: Truss.Compiler.Generate/Syntax/GenerateSyntax.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using Truss.Compiler.Generate.Syntax.Xml;

namespace Truss.Compiler.Generate.Syntax {
    public class GenerateSyntax {
        private readonly Dictionary<string, SyntaxClass> _classes = new Dictionary<string, SyntaxClass>();

        public void Generate(string outputDirectory) {
            var schema = XmlSerialization.Deserialize<SyntaxSchema>(
                Path.Combine(outputDirectory, "Syntax", "SyntaxSchema.xml")
                );

            schema.Classes.Sort((a, b) => a.Name.CompareTo(b.Name));

            foreach (var syntaxClass in schema.Classes) {
                _classes.Add(syntaxClass.Name, syntaxClass);
            }

            var cw = new CodeWriter();

            cw.WriteLine("using System;");
            cw.WriteLine("using System.Collections.Generic;");
            cw.WriteLine("using System.Linq;");
            cw.WriteLine("using System.Text;");
            cw.WriteLine("using Truss.Compiler.Support;");

            cw.WriteLine();
            cw.WriteLine("namespace {0} {{", schema.PackageName);

            cw.Indent();

            foreach (var syntaxClass in schema.Classes) {
                if (!syntaxClass.IsIgnore) {
                    GenerateClass(cw, syntaxClass, schema);
                    cw.WriteLine();
                }
            }

            foreach (var syntaxEnum in schema.Enums) {
                GenerateEnum(cw, syntaxEnum);
                cw.WriteLine();
            }

            GenerateVisitor(schema, cw);
            cw.WriteLine();
            GenerateActionVisitor(schema, cw);
            cw.WriteLine();
            GenerateAbstractVisitor(schema, cw);
            cw.WriteLine();
            GenerateAbstractActionVisitor(schema, cw);
            cw.WriteLine();
            GenerateWalker(schema, cw);
            cw.WriteLine();
            GenerateSyntaxKind(schema, cw);

            cw.UnIndent();

            cw.WriteLine("}");

            WriteWhenChanged(Path.Combine(outputDirectory, "Syntax", "SyntaxSchema.Generated.cs"), cw.ToString());
        }

        private void GenerateVisitor(SyntaxSchema schema, CodeWriter cw) {
            cw.WriteLine("public interface ISyntaxVisitor {");
            cw.Indent();

            cw.WriteLine("bool Done { get; }");

            foreach (var syntaxClass in schema.Classes) {
                if (syntaxClass.IsAbstract) {
                    continue;
                }

                cw.WriteLine();
                cw.WriteLine("void Visit{0}({1} syntax);", GetStrippedName(syntaxClass), syntaxClass.Name);
            }

            cw.UnIndent();
            cw.WriteLine("}");
        }

        private void GenerateActionVisitor(SyntaxSchema schema, CodeWriter cw) {
            cw.WriteLine("public interface ISyntaxVisitor<T> {");
            cw.Indent();

            bool hadOne = false;

            foreach (var syntaxClass in schema.Classes) {
                if (syntaxClass.IsAbstract) {
                    continue;
                }

                if (hadOne) {
                    cw.WriteLine();
                } else {
                    hadOne = true;
                }
                cw.WriteLine("T Visit{0}({1} syntax);", GetStrippedName(syntaxClass), syntaxClass.Name);
            }

            cw.UnIndent();
            cw.WriteLine("}");
        }

        private void GenerateAbstractVisitor(SyntaxSchema schema, CodeWriter cw) {
            cw.WriteLine("public class AbstractSyntaxVisitor : ISyntaxVisitor {");
            cw.Indent();

            cw.WriteLine("public bool Done { get; set; }");
            cw.WriteLine();
            cw.WriteLine("public void DefaultVisit(SyntaxNode syntax) {");
            cw.WriteLine("}");

            foreach (var syntaxClass in schema.Classes) {
                if (syntaxClass.IsAbstract) {
                    continue;
                }

                cw.WriteLine();
                cw.WriteLine("public void Visit{0}({1} syntax) {{", GetStrippedName(syntaxClass), syntaxClass.Name);
                cw.Indent();
                cw.WriteLine("DefaultVisit(syntax);");
                cw.UnIndent();
                cw.WriteLine("}");
            }

            cw.UnIndent();
            cw.WriteLine("}");
        }

        private void GenerateAbstractActionVisitor(SyntaxSchema schema, CodeWriter cw) {
            cw.WriteLine("public class AbstractSyntaxVisitor<T> : ISyntaxVisitor<T> {");
            cw.Indent();

            cw.WriteLine("public T DefaultVisit(SyntaxNode syntax) {");
            cw.Indent();
            cw.WriteLine("return default(T);");
            cw.UnIndent();
            cw.WriteLine("}");

            bool hadOne = false;

            foreach (var syntaxClass in schema.Classes) {
                if (syntaxClass.IsAbstract) {
                    continue;
                }

                if (hadOne) {
                    cw.WriteLine();
                } else {
                    hadOne = true;
                }
                cw.WriteLine("public T Visit{0}({1} syntax) {{", GetStrippedName(syntaxClass), syntaxClass.Name);
                cw.Indent();
                cw.WriteLine("return DefaultVisit(syntax);");
                cw.UnIndent();
                cw.WriteLine("}");
            }

            cw.UnIndent();
            cw.WriteLine("}");
        }

        private void GenerateWalker(SyntaxSchema schema, CodeWriter cw) {
            cw.WriteLine("public class SyntaxTreeWalker : ISyntaxVisitor {");
            cw.Indent();

            cw.WriteLine("public bool Done { get; set; }");
            cw.WriteLine();
            cw.WriteLine("public virtual void VisitList<T>(ImmutableArray<T> list)");
            cw.Indent();
            cw.WriteLine("where T : SyntaxNode {");
            cw.WriteLine("foreach (var node in list) {");
            cw.Indent();
            cw.WriteLine("Visit(node);");
            cw.UnIndent();
            cw.WriteLine("}");
            cw.UnIndent();
            cw.WriteLine("}");
            cw.WriteLine();
            cw.WriteLine("public virtual void Visit(SyntaxNode syntax) {");
            cw.Indent();
            cw.WriteLine("if (syntax != null) {");
            cw.Indent();
            cw.WriteLine("syntax.Accept(this);");
            cw.UnIndent();
            cw.WriteLine("}");
            cw.UnIndent();
            cw.WriteLine("}");

            foreach (var syntaxClass in schema.Classes) {
                if (syntaxClass.IsAbstract) {
                    continue;
                }

                cw.WriteLine();
                cw.WriteLine("public virtual void Visit{0}({1} syntax) {{", GetStrippedName(syntaxClass), syntaxClass.Name);
                cw.Indent();

                foreach (var property in ResolveParameters(syntaxClass)) {
                    if (!IsSyntaxNode(property)) {
                        continue;
                    }

                    if (property.IsList) {
                        cw.WriteLine("VisitList(syntax.{0});", property.Name);
                    } else {
                        cw.WriteLine("Visit(syntax.{0});", property.Name);
                    }
                }

                cw.UnIndent();
                cw.WriteLine("}");
            }

            cw.UnIndent();
            cw.WriteLine("}");
        }

        private void GenerateSyntaxKind(SyntaxSchema schema, CodeWriter cw) {
            cw.WriteLine("public enum SyntaxKind {");
            cw.Indent();

            foreach (var syntaxClass in schema.Classes) {
                if (!syntaxClass.IsAbstract) {
                    cw.WriteLine("{0},", GetStrippedName(syntaxClass));
                }
            }

            cw.UnIndent();
            cw.WriteLine("}");
        }

        private static bool IsSyntaxNode(SyntaxProperty property) {
            return property.Type.EndsWith("Syntax");
        }

        private void GenerateClass(CodeWriter cw, SyntaxClass schemaClass, SyntaxSchema schema) {
            cw.WriteLine(
                "public {0}class {1} : {2} {{",
                schemaClass.IsAbstract ? "abstract " : "",
                schemaClass.Name,
                schemaClass.Base
                );

            cw.Indent();

            var parameters = ResolveParameters(schemaClass);

            cw.Write(
                "{0} {1}(",
                schemaClass.IsAbstract ? "protected" : "public",
                schemaClass.Name
            );

            WriteParameters(cw, parameters, schema);

            cw.WriteLine(")");

            cw.Indent();

            cw.Write(": base(");

            bool hadOne = false;
            foreach (var parameter in parameters) {
                if (schemaClass.Properties.IndexOf(parameter) != -1) {
                    continue;
                }

                if (hadOne) {
                    cw.Write(", ");
                } else {
                    hadOne = true;
                }

                cw.Write(GetLocalName(parameter));
            }

            cw.WriteLine(") {");

            foreach (var parameter in parameters) {
                if (schemaClass.Properties.IndexOf(parameter) == -1) {
                    continue;
                }

                if (!parameter.IsNullable && !IsValueType(parameter.Type, schema)) {
                    cw.WriteLine("if ({0} == null) {{", GetLocalName(parameter));
                    cw.Indent();
                    cw.WriteLine("throw new ArgumentNullException(\"{0}\");", GetLocalName(parameter));
                    cw.UnIndent();
                    cw.WriteLine("}");
                }
            }

            if (schemaClass.Validation != null) {
                cw.WriteRaw(schemaClass.Validation);
                cw.WriteRaw(Environment.NewLine);
            }

            cw.WriteLine();

            foreach (var property in schemaClass.Properties) {
                cw.WriteLine("{0} = {1};", property.Name, GetLocalName(property));
            }

            cw.UnIndent();
            cw.WriteLine("}");

            foreach (var property in schemaClass.Properties) {
                cw.WriteLine();
                cw.WriteLine("public {0} {1} {{ get; private set; }}", GetTypeName(property, schema), property.Name);
            }

            if (!schemaClass.IsAbstract) {
                cw.WriteLine();
                cw.WriteLine("public override SyntaxKind Kind {");
                cw.Indent();
                cw.WriteLine("get {{ return SyntaxKind.{0}; }}", GetStrippedName(schemaClass));
                cw.UnIndent();
                cw.WriteLine("}");

                cw.WriteLine();
                cw.WriteLine("public override void Accept(ISyntaxVisitor visitor) {");
                cw.Indent();
                cw.WriteLine("if (!visitor.Done) {");
                cw.Indent();
                cw.WriteLine("visitor.Visit{0}(this);", GetStrippedName(schemaClass));
                cw.UnIndent();
                cw.WriteLine("}");
                cw.UnIndent();
                cw.WriteLine("}");

                cw.WriteLine();
                cw.WriteLine("public override T Accept<T>(ISyntaxVisitor<T> visitor) {");
                cw.Indent();
                cw.WriteLine("return visitor.Visit{0}(this);", GetStrippedName(schemaClass));
                cw.UnIndent();
                cw.WriteLine("}");
            }

            if (schemaClass.Members != null) {
                cw.WriteRaw(schemaClass.Members);
                cw.WriteRaw(Environment.NewLine);
            }

            cw.UnIndent();
            cw.WriteLine("}");
        }

        private void GenerateEnum(CodeWriter cw, SyntaxEnum schemaEnum) {
            cw.WriteLine(
                "public enum {0} {{",
                schemaEnum.Name
                );

            cw.Indent();

            foreach (var property in schemaEnum.Properties) {
                cw.WriteLine("{0},", property.Name);
            }

            cw.UnIndent();
            cw.WriteLine("}");
        }

        private static void WriteParameters(CodeWriter cw, List<SyntaxProperty> parameters, SyntaxSchema schema) {
            bool hadOne = false;

            foreach (var parameter in parameters) {
                if (hadOne) {
                    cw.Write(", ");
                } else {
                    hadOne = true;
                }

                cw.Write("{0} {1}", GetTypeName(parameter, schema), GetLocalName(parameter));
            }
        }

        private static string GetStrippedName(SyntaxClass syntaxClass) {
            string postfix = "Syntax";
            Debug.Assert(syntaxClass.Name.EndsWith(postfix));
            return syntaxClass.Name.Substring(0, syntaxClass.Name.Length - postfix.Length);
        }

        private static bool IsValueType(string type, SyntaxSchema schema) {
            return type == "bool" || schema.Enums.Any(p => p.Name == type);
        }

        private static string GetLocalName(SyntaxProperty property) {
            string name = property.Name;

            name = name.Substring(0, 1).ToLower() + name.Substring(1);

            switch (name) {
                case "finally":
                case "operator":
                case "else":
                case "default":
                    name = name + "_";
                    break;
            }

            return name;
        }

        private static string GetTypeName(SyntaxProperty property, SyntaxSchema schema) {
            if (property.IsNullable && schema.Enums.Any(p => p.Name == property.Type)) {
                return property.Type + "?";
            }
            if (property.IsList) {
                return "ImmutableArray<" + property.Type + ">";
            }

            return property.Type;
        }

        private List<SyntaxProperty> ResolveParameters(SyntaxClass syntaxClass) {
            List<SyntaxProperty> baseParameters;

            if (syntaxClass.Base != null) {
                baseParameters = ResolveParameters(_classes[syntaxClass.Base]);
            } else {
                baseParameters = new List<SyntaxProperty>();
            }

            var parameters = new List<SyntaxProperty>();
            bool hadLast = false;

            foreach (var baseParameter in baseParameters) {
                if (!hadLast && baseParameter.IsLast) {
                    hadLast = true;
                    parameters.AddRange(syntaxClass.Properties);
                }

                parameters.Add(baseParameter);
            }

            if (!hadLast) {
                parameters.AddRange(syntaxClass.Properties);
            }

            return parameters;
        }

        private static void WriteWhenChanged(string file, string content) {
            if (File.Exists(file)) {
                string current = File.ReadAllText(file);
                if (current == content) {
                    return;
                }
            }

            Directory.CreateDirectory(Path.GetDirectoryName(file));

            File.WriteAllText(file, content);
        }
    }
}
